op-e2e/actions/l1_replica_test.go

```diff
@@ -0,0 +1,102 @@
+package actions
+
+import (
+	"testing"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/consensus/beacon"
+	"github.com/ethereum/go-ethereum/consensus/ethash"
+	"github.com/ethereum/go-ethereum/core"
+	"github.com/ethereum/go-ethereum/core/rawdb"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/stretchr/testify/require"
+
+	"github.com/ethereum-optimism/optimism/op-e2e/e2eutils"
+	"github.com/ethereum-optimism/optimism/op-node/eth"
+	"github.com/ethereum-optimism/optimism/op-node/sources"
+	"github.com/ethereum-optimism/optimism/op-node/testlog"
+)
+
+var defaultRollupTestParams = &e2eutils.TestParams{
+	MaxSequencerDrift:   40,
+	SequencerWindowSize: 120,
+	ChannelTimeout:      120,
+}
+
+var defaultAlloc = &e2eutils.AllocParams{PrefundTestUsers: true}
+
+// Test if we can mock an RPC failure
+func TestL1Replica_ActL1RPCFail(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	dp := e2eutils.MakeDeployParams(t, defaultRollupTestParams)
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+	log := testlog.Logger(t, log.LvlDebug)
+	replica := NewL1Replica(log, sd.L1Cfg)
+	// mock an RPC failure
+	replica.ActL1RPCFail(t)
+	// check RPC failure
+	l1Cl, err := sources.NewL1Client(replica.RPCClient(), log, nil, sources.L1ClientDefaultConfig(sd.RollupCfg, false))
+	require.NoError(t, err)
+	_, err = l1Cl.InfoByLabel(t.Ctx(), eth.Unsafe)
+	require.ErrorContains(t, err, "mock")
+	head, err := l1Cl.InfoByLabel(t.Ctx(), eth.Unsafe)
+	require.NoError(t, err)
+	require.Equal(gt, sd.L1Cfg.ToBlock().Hash(), head.Hash(), "expecting replica to start at genesis")
+}
+
+// Test if we can make the replica sync an artificial L1 chain, rewind it, and reorg it
+func TestL1Replica_ActL1Sync(gt *testing.T) {
+	t := NewDefaultTesting(gt)
+	dp := e2eutils.MakeDeployParams(t, defaultRollupTestParams)
+	sd := e2eutils.Setup(t, dp, defaultAlloc)
+	log := testlog.Logger(t, log.LvlDebug)
+	genesisBlock := sd.L1Cfg.ToBlock()
+	consensus := beacon.New(ethash.NewFaker())
+	db := rawdb.NewMemoryDatabase()
+	sd.L1Cfg.MustCommit(db)
+
+	chainA, _ := core.GenerateChain(sd.L1Cfg.Config, genesisBlock, consensus, db, 10, func(n int, g *core.BlockGen) {
+		g.SetCoinbase(common.Address{'A'})
+	})
+	chainA = append(append([]*types.Block{}, genesisBlock), chainA...)
+	chainB, _ := core.GenerateChain(sd.L1Cfg.Config, chainA[3], consensus, db, 10, func(n int, g *core.BlockGen) {
+		g.SetCoinbase(common.Address{'B'})
+	})
+	chainB = append(append([]*types.Block{}, chainA[:4]...), chainB...)
```

Comment:
does it not matter that these chains are not contiguous?

Reply:
chainB starts on with chainA block number 3 as parent, so the first 4 blocks are common

Reply:
oh I see, nice.